### ISSUE TITLE
feature: add database migrations

### ISSUE THREAD

**Issue by peterschutt** [topsport-com-au/starlite-saqlalchemy#221]
Should the library provide structure for generating database migrations.

If an app uses a database I'm always adding alembic into the downstream environment - so I think it makes sense to include that here.

Should it be another optional feature? Do we simply rely on the alembic cli itself, or add our own cli tool?

**Comment by gazorby** [comment]
> If an app uses a database I'm always adding alembic into the downstream environment - so I think it makes sense to include that here.

Fully aggree

> Should it be another optional feature?

Maybe it should comes with `starlite_saqlalchemy[sqlalchemy]` when #213 is merged?

> Do we simply rely on the alembic cli itself, or add our own cli tool?

If starlite is adding their cli as well, it could be nice if there were some sort plugin support so we could register our alembic commands and everything would be centralized, pretty much the same way Django does with `manage.py` and management commands.

Concerning alembic, I think it could be nice to have a ready to use alembic env and folder structure in one go, something a `"migration init"`. The generated folder structure could looks like that:

```console
...
alembic.ini
alembic/
    env.py
    migrations/
    script.py.mako
```

Basically what it should do is an `alembic init` with a customized `alembic.ini` file and an `env.py` only importing env from the lib:

```python
# alembic/env.py

# Imported env takes care of linking sqlalchemy metadata and having async migrations working
from starlite_saqlalchemy.alembic import env

```

Aso for custom commands, everything you can run with the alembic cli is doable programmatically ircc, so we could add simple abstractions like `migration init`, `migration generate`, `migration apply` `migration unapply -1` etc.

```python
# my_project/alembic/env.py

from starlite_saqlalchemy.alembic import env
```

I actually wrote a CLI wrapping the alembic one some time ago, that setup an alembic env like described above, among other things, so I can help with that if you want.

**Comment by cofin** [comment]
I have several alembic management commands I could contribute here as well.  One thing that I haven't created that would be very nice is a proper `squash migrations` command.  This would function similar to the Django command and allow you to flatten multiple migrations into one.

**Comment by peterschutt** [comment]
@cofin As much as it makes sense, I'd be very happy for this to be as you have it in full-stack-ex. I'll get a PR started, cheers!

**Comment by peterschutt** [comment]
> If starlite is adding their cli as well, it could be nice if there were some sort plugin support so we could register our alembic commands and everything would be centralized, pretty much the same way Django does with manage.py and management commands.

We could go this route as starlite is using click, however, as I'm thinking of supporting a "worker-only" pattern (i.e., starlite would even be optional) so I think we should just have our own cli at the top level.
